src/js/components/CollapsingString.js

```diff
@@ -0,0 +1,161 @@
+import DeepEqual from 'deep-equal';
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import DOMUtils from '../utils/DOMUtils';
+
+const METHODS_TO_BIND = ['updateDimensions'];
+
+class CollapsingString extends React.Component {
+  constructor() {
+    super(...arguments);
+
+    this.state = {
+      collapsed: false,
+      parentWidth: null,
+      stringWidth: null
+    };
+
+    METHODS_TO_BIND.forEach((method) => {
+      this[method] = this[method].bind(this);
+    });
+  }
+
+  componentDidMount() {
+    if (global.window != null) {
+      window.addEventListener('resize', this.updateDimensions);
+      window.addEventListener('focus', this.updateDimensions);
+    }
+
+    this.updateDimensions();
+  }
+
+  componentDidUpdate() {
+    this.updateDimensions();
+  }
+
+  componentWillUnmount() {
+    if (global.window != null) {
+      window.removeEventListener('resize', this.updateDimensions);
+      window.removeEventListener('focus', this.updateDimensions);
+    }
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    if (nextState.parentWidth === null) {
+      return false;
+    }
+
+    return !DeepEqual(this.state, nextState)
+      || !DeepEqual(this.props, nextProps);
+  }
+
+  getParentWidth() {
+    let parent = null;
+    let node = ReactDOM.findDOMNode(this);
+
+    if (this.props.parentSelector != null) {
+      parent = DOMUtils.closest(node, this.props.parentSelector);
+    } else {
+      parent = node.parentNode;
+    }
+
+    return DOMUtils.getComputedWidth(parent);
```

Comment:
Good call, but I think we should default to `document.body` if no parent node is found